src/lazy/preload/index.ts

```diff
@@ -10,74 +10,136 @@ import { insertLinkTag } from './utils';
     typeof __webpack_get_script_filename__ === 'undefined'
   )
-    return false;
+    throw new Error('Unsupported preload strategy');
+
+  return () => {
+    // Replace requireEnsure to create link tags instead of scripts
+    const requireEnsure = __webpack_require__.e;
+    const cleanupLinkTags: Cleanup[] = [];
+    __webpack_require__.e = function requirePreload(chunkId: string) {
+      const href = __webpack_get_script_filename__(chunkId);
+      cleanupLinkTags.push(insertLinkTag(href, rel));
+
+      return fakePromise;
+    };
 
-  // replace requireEnsure to create link tags instead of scripts
-  const requireEnsure = __webpack_require__.e;
-  __webpack_require__.e = function requirePreload(chunkId: string) {
-    const href = __webpack_get_script_filename__(chunkId);
-    insertLinkTag(href, rel);
+    try {
+      loader();
+    } catch (err) {
+      // Ignore any errors
+    }
 
-    return fakePromise;
+    // Restore real webpack require ensure
+    __webpack_require__.e = requireEnsure;
+
+    return () => {
+      for (const cleanupLinkTag of cleanupLinkTags) {
+        cleanupLinkTag();
+      }
+    };
   };
+}
 
-  try {
+export function createLoaderPreloadStrategy({
+  loader,
+}: Pick<PreloadStrategyOptions, 'loader'>): PreloadStrategy {
+  return () => {
     loader();
-  } catch (err) {
-    // ignore
-  }
-  // restore real webpack require ensure
-  __webpack_require__.e = requireEnsure;
 
-  return true;
+    return () => {
+      // Nothing to cleanup...
+    };
+  };
 }
 
-export function preloadAssetViaLoader(loader: Loader<unknown>) {
-  loader();
-
-  return true;
+function isPresent<T>(t: T | undefined | null | void): t is T {
+  return t !== undefined && t !== null;
 }
 
-export function preloadAsset(
-  loader: Loader<unknown>,
-  { moduleId, priority }: { moduleId: string; priority?: PreloadPriority }
-) {
-  if (isNodeEnvironment()) return;
+export type PreloadAssetOptions = {
+  loader: Loader<unknown>;
+  moduleId: string;
+  priority?: PreloadPriority;
+};
+
+export function preloadAsset({
+  loader,
+  moduleId,
+  priority,
+}: PreloadAssetOptions): Cleanup {
+  if (isNodeEnvironment())
+    return () => {
+      // Nothing to cleanup...
+    };
+
   const rel = priority === PRIORITY.HIGH ? 'preload' : 'prefetch';
-  [
-    preloadAssetViaManifest,
-    preloadAssetViaWebpack,
-    preloadAssetViaLoader,
-  ].some(strategy => strategy(loader, { moduleId, rel }));
+  const preloadStrategies = [
+    createManifestPreloadStrategy,
+    createWebpackPreloadStrategy,
+    createLoaderPreloadStrategy,
+  ]
+    .map(strategyFactory => {
+      try {
+        return strategyFactory({ loader, moduleId, rel });
+      } catch (_) {
+        return;
+      }
+    })
+    .filter(isPresent);
+
+  const noopPreloadStrategy = () => () => {
+    // Nothing to cleanup...
+  };
+
+  const preloadStrategy = preloadStrategies[0] ?? noopPreloadStrategy;
+  const cleanupPreload = preloadStrategy();
```

Comment:
can we just:

```
return preloadStrategy()
```

Reply:
Yes, technically. At the time I figured that having the variable would make it a bit clearer to read.